### ISSUE TITLE
Entering/exiting market when hitting target spread without trailing

### ISSUE THREAD

**Issue by alienss** [butor/blackbird#259]
Hey,

Seems that setting TrailingSpreadLim=0.0000 and TrailingSpreadCount=0 doesn't have much of an effect (I was expecting trailing mode to be disabled, and entering/exit the market to be done immediately when hitting or going over the target):

[ 08/12/2017 20:24:30 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,802.9000 / 3,806.2000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,840.9010 / 3,844.0000
   ----------------------------
   Bitstamp/Bitfinex:	-0.28% [target -0.31%, min -0.28%, max  1.72%]

[ 08/12/2017 20:24:33 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,797.8000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,840.9010 / 3,844.0000
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]

[ 08/12/2017 20:24:36 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,797.8000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,835.0510 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:39 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,797.8000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,835.0510 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:42 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,799.8000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,834.1000 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:45 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,799.8000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,834.1000 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:48 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,799.9000 / 3,800.0000
   Bitstamp: 	3,817.0000 / 3,819.8300
   Kraken: 	3,834.1000 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	-0.45% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:51 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,799.9000 / 3,800.0000
   Bitstamp: 	3,782.0100 / 3,801.6700
   Kraken: 	3,834.1000 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	 0.48% [target -0.31%, min -0.45%, max  1.72%]   trailing -0.45%  0/0

[ 08/12/2017 20:24:54 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	3,799.9000 / 3,800.0000
   Bitstamp: 	3,782.0100 / 3,801.6700
   Kraken: 	3,835.0520 / 3,835.6970
   ----------------------------
   Bitstamp/Bitfinex:	 0.48% [target -0.31%, min -0.45%, max  1.72%] 

Basically there were 7 window opportunities to exit the market, yet none were used. Any clue why?

**Comment by bertrandfalguiere** [comment]
It might be because there wasn't enough liquidity in the exchange. 
If you leave the OrderBookFactor parameter untouched, Blackbird wait for the Order book to have 3 times more order volume available to avoid slippage. (see Readme.md)

You can try with a smaller OrderBookFactor, or with smaller volume (TestedExposure) so the factor is more easy to reach for the market, and see if it avoid the problem.

**Comment by alienss** [comment]
Hm, correct me if I'm wrong, but if there's an issue with liquidity, there's usually a message specifying that, something like
<pre>
WARNING: Opportunity found but not enough liquidity. Trade canceled
</pre>

Are there situations where there might be liquidity issues, but the message wouldn't show up (like in my initial post)?

**Comment by bertrandfalguiere** [comment]
I'm new to the project and trade small amount, so I never had a liquidity problem and didn't new the warning existed.

The spread is not the only number staying the same: all bid and ask prices remain the same, which is very surprising for exchanges of this size. Blackbird may have made the right decision based on the data it have, but these seems to be incorrect. Connection issue? Internal bug? Unable to handle zero spread? I'm not familiar enough with the code.

**Comment by alienss** [comment]
Also trading small amount (25$), so doubt it's a liquidity issue. Doubt there's a problem with the data, it's not highly surprising for the prices to remain the same for 10-15 seconds.

Thing is I'd like to know how to force blackbird to enter/exit without using trailing, to check if there are more opportunities being used for entering/exiting (especially exiting).

**Comment by alienss** [comment]
Here's another example where trailing can mean a lost exit opportunity, which means your balance is still in the market, locked out of any future spread targets:

<pre>
[ 08/15/2017 12:57:51 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,060.1000 / 4,065.6000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,109.1000 / 4,109.5130
   ----------------------------
   Bitstamp/Bitfinex:	-0.33% [target -0.50%, min -1.42%, max  1.77%]

[ 08/15/2017 12:57:54 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,109.1000 / 4,109.3100
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]

[ 08/15/2017 12:57:57 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,109.1000 / 4,109.3100
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:00 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,109.1000 / 4,109.5130
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:03 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,109.1000 / 4,109.5130
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:06 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,110.0000 / 4,111.9910
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:09 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,079.0000 / 4,083.6900
   Kraken: 	4,110.0000 / 4,111.9910
   ----------------------------
   Bitstamp/Bitfinex:	-0.63% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:12 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,066.6600 / 4,074.5300
   Kraken: 	4,110.0000 / 4,111.9910
   ----------------------------
   Bitstamp/Bitfinex:	-0.33% [target -0.50%, min -1.42%, max  1.77%]   trailing -0.58%  0/1

[ 08/15/2017 12:58:15 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,066.6600 / 4,074.5300
   Kraken: 	4,110.0000 / 4,111.9910
   ----------------------------
   Bitstamp/Bitfinex:	-0.33% [target -0.50%, min -1.42%, max  1.77%]

[ 08/15/2017 12:58:18 IN MARKET: Long Bitstamp / Short Bitfinex ]
   Bitfinex: 	4,053.3000 / 4,053.4000
   Bitstamp: 	4,066.6600 / 4,074.5300
   Kraken: 	4,110.0000 / 4,111.9910
   ----------------------------
   Bitstamp/Bitfinex:	-0.33% [target -0.50%, min -1.42%, max  1.77%]
</pre>

**Comment by butor** [comment]
@alienss Yes we should add something like:

_if params.trailingCount == 0 then disable trailing_

**Comment by alienss** [comment]
Hey butor,

That would be awesome.

I also noticed two odd things with res.trailing in check_entry_exit.cpp:

**1. Defining res.trailing in checkEntry():**
https://github.com/butor/blackbird/blob/4769bec94cc7e7238f67ed2af03e2bdcc68685be/src/check_entry_exit.cpp#L94-L98

What happens if I start blackbird with a spreadEntry of 0.50%, and in the first iteration the spreadIn is 0.80%? Apparently, **res.trailing** doesn't get set to -1.0, since res.spreadIn > params.spreadEntry right away. That also means that if `(res.trailing[longId][shortId] == -1.0)` won't be executed, so when we get to `if (newTrailValue >= res.trailing[longId][shortId])`, what will be the value of res.trailing?

Perhaps res.trailing should be defined in main.cpp as well, when blackbird starts, before the main analysis loop (and before we load restore.txt values, just in case we're in market when launching).

**2. Extra loop when calculating res.trailing (both for checkEntry and checkExit):** 
https://github.com/butor/blackbird/blob/4769bec94cc7e7238f67ed2af03e2bdcc68685be/src/check_entry_exit.cpp#L184-L187

So with a checkExit example, if res.spreadOut < res.exitTarget (which means that there should be an exit opportunity), we define res.trailing and then `return false`. That means that we can only take advantage of the opportunity from the next loop going forward.

Let's asume we're **IN MARKET** and have a spreadExit of **-0.50%**:

| Spread Ratio | Action |
| --- | --- |
| -0.40% | res.trailing is reset, return false (exit the loop) |
| -0.40% | res.trailing is reset, return false (exit the loop) |
| -0.55% | **set res.trailing, but then return false (exit the loop)** |
| -0.55% | do the checks to see if there's an opportunity |

Let me know if that makes sense. I don't know much C++, so it's very possible I might be wrong here.

Cheers!

**Comment by greatwolf** [comment]
`result` is initialized in `reset` in the reset method: https://github.com/butor/blackbird/blob/4769bec94cc7e7238f67ed2af03e2bdcc68685be/src/result.cpp#L61-L72

and main does call it prior to use: https://github.com/butor/blackbird/blob/4769bec94cc7e7238f67ed2af03e2bdcc68685be/src/main.cpp#L335-L336

That said, this should probably be done in the constructor so the usage is less awkward. I didn't do that in the refactor since I didn't want to change the code too much. I'm sure @butor would welcome a PR for this though.

Not sure what the intended behavior of trailing mechanism should be though.

**Comment by alienss** [comment]
Ah ok, didn't check result.cpp.

**Comment by butor** [comment]
Any quality PR is welcome ;-)

**Comment by alienss** [comment]
> 2. Extra loop when calculating res.trailing (both for checkEntry and checkExit):

I rechecked the logic, if trailing is enabled, then returning false there is ok, since you can't do anything else in the first iteration anyway (since you're waiting for the second one in order to spread goes higher or lower).

**Comment by nznelson** [comment]
Hi all, just wondering, did this ever get resolved? Is it possible to set TrailingSpreadLim & TrailingSpreadCount to 0 to disable trailing?
